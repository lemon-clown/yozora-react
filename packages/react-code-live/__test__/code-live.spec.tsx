import { jest } from '@jest/globals'
import { fireEvent, render, waitFor } from '@testing-library/react'
import CodeRendererJsx from '@yozora/react-code-renderer-jsx'
import type { ICodeRunnerItem, ICodeRunnerProps } from '@yozora/react-code-runners'
import React from 'react'
import CodeLive from '../src'

const code = `
  function Counter() {
    const [count, setCount] = React.useState(0)
    return (
      <div>
        <button onClick={() => setCount(c => c + 1)}>+</button>
        <span data-testid="value">{count}</span>
        <button onClick={() => setCount(c => c - 1)}>-</button>
      </div>
    )
  }
`

const JsxRunner: React.FC<ICodeRunnerProps> = ({ value, onError }) => {
  return <CodeRendererJsx code={value} inline={true} onError={onError} />
}

const runners: ICodeRunnerItem[] = [
  {
    title: 'jsx',
    pattern: /^jsx$/,
    runner: JsxRunner,
  },
]

// Fake timers using Jest
beforeEach(() => {
  jest.useFakeTimers()
})

// Running all pending timers and switching to real timers using Jest
afterEach(() => {
  jest.runOnlyPendingTimers()
  jest.useRealTimers()
})

describe('prop types', () => {
  test('change and debounce', async () => {
    const code1 = 'function Demo() { return <span data-testid="value">3</span> }'
    const code2 = 'function Demo() { return <span data-testid="value">4</span> }'

    const view = render(<CodeLive lang="jsx" value={code1} runners={runners} />)
    await waitFor(() => {
      const textarea = view.getByRole('textbox')
      expect(textarea.textContent).toEqual(code1)
      expect(view.getByTestId('value').textContent).toEqual('3')
    })

    // change code
    fireEvent.change(view.getByRole('textbox'), { target: { value: code2 } })

    // Fast-forward time (await debounce)
    await waitFor(() => {
      expect(view.getByRole('textbox').textContent).toEqual(code2)
      expect(view.getByTestId('value').textContent).toEqual('4')
    })
  })
})

describe('snapshot', () => {
  test('basic', () => {
    const view = render(<CodeLive lang="jsx" value={code} runners={runners} />)
    expect(view.getByRole('textbox').textContent).toEqual(code)
    expect(view.getByTestId('value').textContent).toEqual('0')
    expect(view.asFragment()).toMatchSnapshot()
  })
})
