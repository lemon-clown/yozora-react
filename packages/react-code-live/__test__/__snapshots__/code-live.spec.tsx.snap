// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot basic 1`] = `
<DocumentFragment>
  <div
    class="yozora-code-live css-1jh9kfk"
  >
    <div
      class="css-sgdhwv"
    >
      <span
        class="css-9y6172"
      >
        <span
          class="css-t4q37b"
          title="close"
        />
        <span
          class="css-p28dfd"
          title="minimize"
        />
        <span
          class="css-6luv6p"
          title="maximize"
        />
      </span>
      <span
        class="css-1p3aks"
      >
        Â 
      </span>
      <span
        class="css-nsmueo"
      >
        <button
          aria-label="Copy to clipboard"
          class="css-jbwb7c"
          data-copy-status="waiting"
          type="button"
        >
          copy
        </button>
      </span>
    </div>
    <div
      class="css-17fon8l"
    >
      <div
        class="css-1eksy5g"
      >
        <div
          class="yozora-code-editor css-kt1ey1 yozora-code-editor"
          padding="10"
          style="line-height: 1.6rem; tab-size: 2;"
        >
          <div
            class="css-6gcr58"
          >
            <div
              class="css-1tu59u4"
              style="width: 2.2em;"
            />
            <textarea
              autocapitalize="off"
              autocomplete="off"
              autocorrect="off"
              class="css-1pinvdg"
              spellcheck="false"
              style="tab-size: 2; white-space: pre; padding: 0.5rem 0.6em 0px 12px;"
            >
                function Counter() {
    const [count, setCount] = React.useState(0)
    return (
      &lt;div&gt;
        &lt;button onClick={() =&gt; setCount(c =&gt; c + 1)}&gt;+&lt;/button&gt;
        &lt;span data-testid="value"&gt;{count}&lt;/span&gt;
        &lt;button onClick={() =&gt; setCount(c =&gt; c - 1)}&gt;-&lt;/button&gt;
      &lt;/div&gt;
    )
  }

            </textarea>
          </div>
          <pre
            aria-hidden="true"
            class="css-14t25us"
            style="tab-size: 2; white-space: pre;"
          >
            <div
              class="css-1k8659m prism-code language-jsx"
              style="color: rgb(212, 212, 212); line-height: 1.6rem; max-height: 20.8rem; min-height: 100%;"
            >
              <div
                class="css-hfkijx"
                style="width: 2.2em;"
              >
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    1
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    2
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    3
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    4
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    5
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    6
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    7
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    8
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    9
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    10
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    11
                  </span>
                </div>
                <div
                  class="css-17fnzo2"
                >
                  <span>
                    12
                  </span>
                </div>
              </div>
              <div
                class="css-rv5jjl"
              >
                <div
                  class="css-1ttfemb"
                >
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain"
                      style="display: inline-block;"
                    >
                      

                    </span>
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain"
                    >
                        
                    </span>
                    <span
                      class="token keyword"
                      style="color: rgb(86, 156, 214);"
                    >
                      function
                    </span>
                    <span
                      class="token plain"
                    >
                       
                    </span>
                    <span
                      class="token function"
                      style="color: rgb(86, 156, 214);"
                    >
                      Counter
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      (
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      )
                    </span>
                    <span
                      class="token plain"
                    >
                       
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      {
                    </span>
                    <span
                      class="token plain"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain"
                    >
                          
                    </span>
                    <span
                      class="token keyword"
                      style="color: rgb(86, 156, 214);"
                    >
                      const
                    </span>
                    <span
                      class="token plain"
                    >
                       
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      [
                    </span>
                    <span
                      class="token plain"
                    >
                      count
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      ,
                    </span>
                    <span
                      class="token plain"
                    >
                       setCount
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      ]
                    </span>
                    <span
                      class="token plain"
                    >
                       
                    </span>
                    <span
                      class="token operator"
                      style="color: rgb(212, 212, 212);"
                    >
                      =
                    </span>
                    <span
                      class="token plain"
                    >
                       React
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      .
                    </span>
                    <span
                      class="token function"
                      style="color: rgb(86, 156, 214);"
                    >
                      useState
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      (
                    </span>
                    <span
                      class="token number"
                      style="color: rgb(181, 206, 168);"
                    >
                      0
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      )
                    </span>
                    <span
                      class="token plain"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain"
                    >
                          
                    </span>
                    <span
                      class="token keyword"
                      style="color: rgb(86, 156, 214);"
                    >
                      return
                    </span>
                    <span
                      class="token plain"
                    >
                       
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      (
                    </span>
                    <span
                      class="token plain"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain"
                    >
                            
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &lt;
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                      div
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &gt;
                    </span>
                    <span
                      class="token plain-text"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain-text"
                    >
                              
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &lt;
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                      button
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag attr-name"
                      style="color: rgb(156, 220, 254);"
                    >
                      onClick
                    </span>
                    <span
                      class="token tag script language-javascript script-punctuation punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      =
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      {
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      (
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      )
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag script language-javascript operator"
                      style="color: rgb(212, 212, 212);"
                    >
                      =&gt;
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag script language-javascript function"
                      style="color: rgb(86, 156, 214);"
                    >
                      setCount
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      (
                    </span>
                    <span
                      class="token tag script language-javascript parameter"
                      style="color: rgb(86, 156, 214);"
                    >
                      c
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag script language-javascript operator"
                      style="color: rgb(212, 212, 212);"
                    >
                      =&gt;
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       c 
                    </span>
                    <span
                      class="token tag script language-javascript operator"
                      style="color: rgb(212, 212, 212);"
                    >
                      +
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag script language-javascript number"
                      style="color: rgb(181, 206, 168);"
                    >
                      1
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      )
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      }
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &gt;
                    </span>
                    <span
                      class="token plain-text"
                    >
                      +
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &lt;/
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                      button
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &gt;
                    </span>
                    <span
                      class="token plain-text"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain-text"
                    >
                              
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &lt;
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                      span
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag attr-name"
                      style="color: rgb(156, 220, 254);"
                    >
                      data-testid
                    </span>
                    <span
                      class="token tag attr-value punctuation attr-equals"
                      style="color: rgb(106, 153, 85);"
                    >
                      =
                    </span>
                    <span
                      class="token tag attr-value punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      "
                    </span>
                    <span
                      class="token tag attr-value"
                      style="color: rgb(206, 145, 120);"
                    >
                      value
                    </span>
                    <span
                      class="token tag attr-value punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      "
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &gt;
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      {
                    </span>
                    <span
                      class="token plain"
                    >
                      count
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      }
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &lt;/
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                      span
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &gt;
                    </span>
                    <span
                      class="token plain-text"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain-text"
                    >
                              
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &lt;
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                      button
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag attr-name"
                      style="color: rgb(156, 220, 254);"
                    >
                      onClick
                    </span>
                    <span
                      class="token tag script language-javascript script-punctuation punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      =
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      {
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      (
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      )
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag script language-javascript operator"
                      style="color: rgb(212, 212, 212);"
                    >
                      =&gt;
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag script language-javascript function"
                      style="color: rgb(86, 156, 214);"
                    >
                      setCount
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      (
                    </span>
                    <span
                      class="token tag script language-javascript parameter"
                      style="color: rgb(86, 156, 214);"
                    >
                      c
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag script language-javascript operator"
                      style="color: rgb(212, 212, 212);"
                    >
                      =&gt;
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       c 
                    </span>
                    <span
                      class="token tag script language-javascript operator"
                      style="color: rgb(212, 212, 212);"
                    >
                      -
                    </span>
                    <span
                      class="token tag script language-javascript"
                      style="color: rgb(86, 156, 214);"
                    >
                       
                    </span>
                    <span
                      class="token tag script language-javascript number"
                      style="color: rgb(181, 206, 168);"
                    >
                      1
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      )
                    </span>
                    <span
                      class="token tag script language-javascript punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      }
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &gt;
                    </span>
                    <span
                      class="token plain-text"
                    >
                      -
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &lt;/
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                      button
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &gt;
                    </span>
                    <span
                      class="token plain-text"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain-text"
                    >
                            
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &lt;/
                    </span>
                    <span
                      class="token tag"
                      style="color: rgb(86, 156, 214);"
                    >
                      div
                    </span>
                    <span
                      class="token tag punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      &gt;
                    </span>
                    <span
                      class="token plain"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain"
                    >
                          
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      )
                    </span>
                    <span
                      class="token plain"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain"
                    >
                        
                    </span>
                    <span
                      class="token punctuation"
                      style="color: rgb(106, 153, 85);"
                    >
                      }
                    </span>
                    <span
                      class="token plain"
                    />
                  </div>
                  <div
                    class="token-line css-7zjwmr"
                    style="color: rgb(212, 212, 212);"
                  >
                    <span
                      class="token plain"
                      style="display: inline-block;"
                    >
                      

                    </span>
                  </div>
                </div>
              </div>
            </div>
          </pre>
        </div>
      </div>
      <div
        class="css-ov3fus"
      >
        <div
          class="yozora-code-embed css-1lkb549"
        >
          <div>
            <button>
              +
            </button>
            <span
              data-testid="value"
            >
              0
            </span>
            <button>
              -
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
